### DISCUSSION TITLE
Removing the hash from generated classNames

### DISCUSSION THREAD

**Discussion by jamestowers** [vanilla-extract-css/vanilla-extract#273] [Q&A] [answered]
I appreciate this kind of goes against the whole point of scoped CSS modules but is it possible? 

I've tried to play around with the Webpack css-loader `options.modules`  config but to no avail. 

I'm interested in being able to use the generated classes outside of a JS environment - in server side templates - and the BEM style format you get in dev mode (with webpack) could work well, however obviously the hashes mean the classNames change on every build. 
For context, we're running a React frontend but are exploring rendering some static pages on the server.

**Comment by mattcompiles** [comment, marked as answer]
I have a few of thoughts on this. Firstly I'd note that the classNames don't change on each build, they should be consistent between builds as long as you don't change the file name or re-order the styles. This is designed to ensure long-term caching works as well as possible. 

On your use-case of using the classNames on the server, I'm assuming your server isn't running node? Because if it is you'd be better off importing the `.css.ts` files directly and using them just like you do in React. If you're using a different server setup then I can see why this isn't a simple issue. 

If we removed hashes from the generated classes you'd be extremely likely to cause class collisions. Also the debug id feature (i.e. the thing that names the classes after your variable names) isn't a perfect system and probably shouldn't be relied on to make your production code work.

**Reply by jamestowers** [reply]
Thanks Matt,

That all makes sense, I hadn't noticed that the classname more often remain consistent, that's good to know.

Your assumption is correct, we're not running node on the server so that won't be an option.

The unreliable debug names feels like the last nail in this coffin, it was a far fetched idea but wanted to rule it out, thanks!

**Reply by phil-lgr** [reply]
As others have pointed out, sometimes you have to apply CSS vars and classes to non react or node applications.

In that case, you'll want to generate a final .css or multiples based on type safe vanilla's sprinkles and styles... and simply use the classes.

I think we should have an option to generate sprinkles and style classnames without hash (it has been already done for vars here https://github.com/seek-oss/vanilla-extract/discussions/313) so why not surface an option in the plugins or directly with a function's option?

**Reply by phil-lgr** [reply]
Also requested here: https://github.com/seek-oss/vanilla-extract/discussions/313#discussioncomment-1544009

**Reply by phil-lgr** [reply]
Similarly to what has been added to `createGlobalThemeContract`

![image](https://user-images.githubusercontent.com/9644867/145092903-431c1f3a-2a88-4402-b772-a15f88e548b5.png)

We could have an option for `defineProperties` and `style`?

There are use cases where you will want to generate unique classnames without hashes (e.g. global theming) and export that for non-js environments

**Comment by RyderJKL** [comment]
It is still expected that the ability to remove hash will be supported, because in real business systems（for example, in the ui component library project）, where refactoring is often done, the directory structure and names of components are likely to change and the resulting hash will be disastrous for the real environment. This is because higher level components are likely to overwrite the old classname_hash.

**Comment by adamduncan** [comment]
> If we removed hashes from the generated classes you'd be extremely likely to cause class collisions.

@mattcompiles Agree this certainly introduces that risk. Looking at the possibilities from the perspective of applications that aren't necessarily JS-/Node-based, I wonder whether there's some value in offering the ability to customise classnames.

As an example: One might be designing a component library, knowing components will _mostly_ be used within React applications. However to afford reuse through static HTML sites and, say, Ruby or Django templating, components would benefit from a developer-friendly HTML classname API of some kind. Traditionally we'd leverage a convention (e.g. BEM) to mitigate _some_ of the risk of class collisions, granted it's not bulletproof. E.g. As an example from Braid, we might hope for:

```
<div class="stack stack--align-center">
     <div class="stack__item">...</div>
     <div class="stack__item">...</div>
</div>
```

Does that seem like something that could be a consideration of `vanilla-extract` in future? You mention it being not such a simple issue for non-JS environments. Do you have any thoughts on how one might faciliate this reuse, while still reaping the authoring/type-safety benefits `vanilla-extract` brings?

Any input you have would be much appreciated. Cheers 👍 

Edit: Related to https://github.com/seek-oss/vanilla-extract/discussions/234#discussioncomment-1016987

**Reply by mryechkin** [reply]
Has anyone found a solution or workaround for this? Currently investigating `vanilla-extract` to replace our existing styling solution in a design system component library we're building, and this is exactly the use case. Most consuming applications will be in React, however there are some cases where apps are simple HTML/CSS, and could benefit from consuming the same styles as the React components from the design system.

@jamestowers did you end up finding a solution for your use case?

**Reply by jamestowers** [reply]
Apologies for the later reply, I only just came back here to see if there was any news on the issue. I stopped looking for a solution but I'm still keen to find one. Passing a custom hashing function as in the linked discussion above could get us a long way there

**Reply by nmkae** [reply]
We're hoping to achieve a similar result where we are hoping to expose some utility sprinkles to support tailwind like class names on a simple div (without needing to import a function or drive it through another component) while also keeping the hashed class names for components (where class collision will be a concern).  It's not clear if this is achievable in the current Vanilla Extract or if it will ever be.

**Comment by askoufis** [comment]
For future reference, fully custom hashes are now possible as of https://github.com/vanilla-extract-css/vanilla-extract/pull/1160.
